test/unit/org/apache/cassandra/cql3/SecondaryIndexSSTableExportTest.java

```diff
@@ -0,0 +1,181 @@
+        String insert = "INSERT INTO %s (k, v, s, c, m, l, st) VALUES (0, 0, 's', 10, {100:'v'}, ['l1', 'l2'], {1, 2, 3})";
+        indexSstableValidation(createTable, createIndex, insert);
+    }
+
+    @Test
+    public void testCollectionListIndex() throws Throwable
+    {
+        String createTable = "CREATE TABLE %s (k int , v int, s  text static, c bigint, m map<bigint, text>, l list<text>, st set<int>, primary key((k, v), c))";
+        String createIndex = "CREATE INDEX ON %s (l)";
+        String insert = "INSERT INTO %s (k, v, s, c, m, l, st) VALUES (0, 0, 's', 10, {100:'v'}, ['l1', 'l2'], {1, 2, 3})";
+        indexSstableValidation(createTable, createIndex, insert);
+    }
+
+    @Test
+    public void testCollectionSetIndex() throws Throwable
+    {
+        String createTable = "CREATE TABLE %s (k int , v int, s  text static, c bigint, m map<bigint, text>, l list<text>, st set<int>, primary key((k, v), c))";
+        String createIndex = "CREATE INDEX ON %s (st)";
+        String insert = "INSERT INTO %s (k, v, s, c, m, l, st) VALUES (0, 0, 's', 10, {100:'v'}, ['l1', 'l2'], {1, 2, 3})";
+        indexSstableValidation(createTable, createIndex, insert);
+    }
+
+    private void indexSstableValidation(String createTableCql, String createIndexCql, String insertCql) throws Throwable
+    {
+        String SUCEESS_MSG = "Test passed";
+        try
+        {
+            Pair<String, String> tableIndex = generateSstable(createTableCql, createIndexCql, insertCql);
+            ColumnFamilyStore cfs = getColumnFamilyStore(KEYSPACE, tableIndex.left);
+            assertTrue(cfs.indexManager.hasIndexes());
+            assertNotNull(cfs.indexManager.getIndexByName(tableIndex.right));
+            for (ColumnFamilyStore columnFamilyStore : cfs.indexManager.getAllIndexColumnFamilyStores())
+            {
+                assertTrue(columnFamilyStore.isIndex());
+                assertFalse(columnFamilyStore.getLiveSSTables().isEmpty());
+                for (SSTableReader sst : columnFamilyStore.getLiveSSTables())
+                {
+                    String file = sst.getFilename();
+                    ToolRunner.ToolResult tool = ToolRunner.invokeClass(SSTableExport.class, file);
+                    List<Map<String, Object>> parsed = JsonUtils.JSON_OBJECT_MAPPER.readValue(tool.getStdout(), jacksonListOfMapsType);
+                    assertNotNull(tool.getStdout(), parsed.get(0).get("partition"));
+                    assertNotNull(tool.getStdout(), parsed.get(0).get("rows"));
+                }
+            }
+            fail(SUCEESS_MSG);
+        }
+        catch (Throwable throwable)
+        {
+            // UPGRADING or NONE
+            if (DatabaseDescriptor.getStorageCompatibilityMode().isAfter(5))
+            {
+                assertEquals(SUCEESS_MSG, throwable.getMessage());
+            }
+            else
+            {
+                System.out.println(throwable.getMessage());
```

Comment:
I think we can check that we are actually throwing the expected message, and mentioning CASSANDRA-18254 since we are not entirely fixing the bug:
```java
Pair<String, String> tableIndex = generateSstable(createTableCql, createIndexCql, insertCql);
ColumnFamilyStore cfs = getColumnFamilyStore(KEYSPACE, tableIndex.left);
assertTrue(cfs.indexManager.hasIndexes());
assertNotNull(cfs.indexManager.getIndexByName(tableIndex.right));
for (ColumnFamilyStore columnFamilyStore : cfs.indexManager.getAllIndexColumnFamilyStores())
{
    assertTrue(columnFamilyStore.isIndex());
    assertFalse(columnFamilyStore.getLiveSSTables().isEmpty());
    for (SSTableReader sst : columnFamilyStore.getLiveSSTables())
    {
        String file = sst.getFilename();
        try
        {
            ToolRunner.ToolResult tool = ToolRunner.invokeClass(SSTableExport.class, file);
            List<Map<String, Object>> parsed = JsonUtils.JSON_OBJECT_MAPPER.readValue(tool.getStdout(), jacksonListOfMapsType);
            assertNotNull(tool.getStdout(), parsed.get(0).get("partition"));
            assertNotNull(tool.getStdout(), parsed.get(0).get("rows"));
        }
        catch (AssertionError e)
        {
            // TODO: CASSANDRA-18254 should provide a workaround for pre-5.0 sstables
            assertTrue(DatabaseDescriptor.getStorageCompatibilityMode().isBefore(5));
            Assertions.assertThat(e.getMessage())
                      .contains("PartitionerDefinedOrder's toJSONString method needs a partition key type but now is null.");
        }
        catch (MismatchedInputException e)
        {
            // TODO: CASSANDRA-18254 should provide a workaround for pre-5.0 sstables
            assertTrue(DatabaseDescriptor.getStorageCompatibilityMode().isBefore(5));
            Assertions.assertThat(e.getMessage())
                      .contains("No content to map due to end-of-input");
        }
    }
}
```

Reply:
ok